go/vt/vtgate/planbuilder/dml.go

```diff
@@ -0,0 +1,204 @@
+/*
+Copyright 2019 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package planbuilder
+
+import (
+	"vitess.io/vitess/go/sqltypes"
+	topodatapb "vitess.io/vitess/go/vt/proto/topodata"
+	vtrpcpb "vitess.io/vitess/go/vt/proto/vtrpc"
+	"vitess.io/vitess/go/vt/sqlparser"
+	"vitess.io/vitess/go/vt/vterrors"
+	"vitess.io/vitess/go/vt/vtgate/engine"
+	"vitess.io/vitess/go/vt/vtgate/vindexes"
+)
+
+// getDMLRouting returns the vindex and values for the DML,
+// If it cannot find a unique vindex match, it returns an error.
+func getDMLRouting(where *sqlparser.Where, table *vindexes.Table) (engine.DMLOpcode, vindexes.SingleColumn, string, []sqltypes.PlanValue) {
+	var keyColumn string
+	for _, index := range table.Ordered {
+		if !index.Vindex.IsUnique() {
+			continue
+		}
+		single, ok := index.Vindex.(vindexes.SingleColumn)
+		if !ok {
+			continue
+		}
+
+		if keyColumn == "" { //TODO - check with sougou on how to deterministically get the primary vindex column
+			keyColumn = sqlparser.String(index.Columns[0])
+		}
+
+		if where == nil {
```

Comment:
This can be moved outside the loop.

Reply:
Well... we need to enter the loop at least once, which we achieve by doing the check in here. We need this to find the keyColumn. Maybe there is a better way of doing it that we are not seeing?